tiledb/cc/vfs.cc

```diff
@@ -41,6 +43,34 @@ void init_vfs(py::module &m) {
       .def("_copy_file", &VFS::copy_file)
 
       .def("_ls", &VFS::ls)
+      .def(
+          "_ls_recursive",
+          // 1. the user provides a callback, the user callback is passed to the
+          // C++ function. Return an empty vector.
+          // 2. no callback is passed and a default callback is used. The
+          // default callback just appends each path gathered. Return the paths.
+          [](VFS &vfs, std::string uri, py::object callback) {
+            tiledb::Context ctx;
+            if (callback.is_none()) {
+              std::vector<std::string> paths;
+              tiledb::VFSExperimental::ls_recursive(
+                  ctx, vfs, uri,
+                  [&](const std::string_view path,
+                      uint64_t object_size) -> bool {
+                    paths.push_back(std::string(path));
+                    return true;
+                  });
+              return paths;
```

Comment:
For awareness, please note that this will be expensive for large result sets, because pybind11 will generate code to convert it to a native py::list (each item has to be copied into the list before returning - see discussion here: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#binding-stl-containers). The only way to avoid that is to a create a holder class like this:
```
class PbStdVectorString {
    std::vector<string> m_vec;
    PbStdVectorString(std::string&& v) : m_vec(v) {}
    ...
```
and then wrap `PbStdVectorString` class in pybind11 to provide getitem/slice/iter for views on the underlying data. Not a blocker but might be necessary when you benchmark.